docs/manual/kernel/Vmm.tex

```diff
@@ -0,0 +1,235 @@
+The VMM and VM drivers make full usage of page faults (and handling them). Often a fault is triggered by the hardware in response to action taken by a program, but a page fault can also be triggered manually by calling \verb|HandleFault()| on any VMM to trigger the same response. This function returns a bool indicating if the fault was considered 'good' or not. A good fault is an allowed (and expected) memory access, while a bad fault is an illegal one.
+
+Bad faults can result in any number of consequences, from a program being terminated to a kernel panic, depending on the VMM in question.
+
+When handling a page fault the VMM will check the faulting address against the list of active VM ranges, looking for the associated range. If a matching range isn't found, the fault returns \verb|false| as it failed to handle it properly. If the range is found, the VMM will then check the fault flags which describe what the memory access was (read, write or execute), it's privilege level (supervisor or user). Assuming the memory access was allowed for this VM range, the VMM will call the fault handler on the VM driver attached to the range and allow the driver to continue handling the fault. At this point the VM driver is responsible for the outcome of the fault.
+
+VM drivers may choose to not map physical memory with the full permissions requested by the VMM, in order to trigger a fault on certain actions. This is used to implement features like CoW (copy-on-write), zero-paging, or to allow the backing memory to be reclaimed and used elsewhere. \textit{Currently swapping to disk isn't implemented, and the file cache is only half transient - it has no page-out features}
+
+At this point the action taken depends on the specific VM driver and the exact nature of the fault.
+
+\subsection{Example Code}
+Allocating and freeing can be done by calling the VMM directly, or by using a VMO. The kernel VMM is accessible through a static function \lstinline{VMM::Kernel()} and the active user VMM can be access similarly as \lstinline{VMM::Current()}. The VMM and associated structs are declared in \verb|memory/Vmm.h|, to use VMOs without including the rest of the virtual memory related code you can just include |verb|memory/VmObject.h|.
+
+The \lstinline{Alloc()} function accepts a size and flags argument. The flags contains permissions (write and execute, all memory is readable), the type of memory required (MMIO, anon or file-backed) and some other flags for adjusting the behaviour of this VM range. There is also an \lstinline{initArg} argument, the purpose of which depends on the type of memory requested: for MMIO it's the base physical address of the MMIO to map, for anonymous memory it's a set of feature flags to disable for this allocation, and for for-backed memory it contains a pointer to a \lstinline{VmoFileInitArg}.
+
+As an example we'll allocate 123 bytes of writable anonymous virtual memory in kernel address space, and then free it.
+\begin{lstlisting}
```

Comment:
again bad rendering of begin, just double check that the pdf is rendered correctly.